test/fixtures/parser/postgres/postgres_create_function.sql

```diff
@@ -8,4 +8,15 @@ CREATE OR REPLACE FUNCTION increment(i integer) RETURNS integer AS '
     BEGIN
         RETURN i + 1;
     END;
-' LANGUAGE plpgsql;
+' LANGUAGE plpgsql VOLATILE;
+
+CREATE OR REPLACE FUNCTION increment(i integer) RETURNS integer AS '
+    BEGIN
+        RETURN i + 1;
+    END;
+' LANGUAGE plpgsql WINDOW IMMUTABLE STABLE LEAKPROOF RETURNS NULL ON NULL INPUT EXTERNAL SECURITY DEFINER
+ROWS 5 SET test_param = 3;
+
+CREATE OR REPLACE FUNCTION increment(i integer) RETURNS integer AS 'C:\\my_file.c', 'symlink_c'
+LANGUAGE plpgsql WINDOW IMMUTABLE STABLE NOT LEAKPROOF CALLED ON NULL INPUT EXTERNAL SECURITY DEFINER COST 123
+ROWS 5 SET test_param = 3 WITH (isStrict);
```

Comment:
It might be good to have some more examples of create function here. Two examples for all that syntax isn't a huge amount to be honest.

Reply:
Added one more example. I can add more if needed, but i think all bases are covered by the tests now

Reply:
What do you think about adding all the examples from the spec as well? So it's not jsut examples we've come up wiht (using whatever assumptions we've made)?

```sql

-- Exmaples from https://www.postgresql.org/docs/13/sql-createfunction.html

CREATE FUNCTION add(integer, integer) RETURNS integer
    AS 'select $1 + $2;'
    LANGUAGE SQL
    IMMUTABLE
    RETURNS NULL ON NULL INPUT;

CREATE OR REPLACE FUNCTION increment(i integer) RETURNS integer AS $$
        BEGIN
                RETURN i + 1;
        END;
$$ LANGUAGE plpgsql;


CREATE FUNCTION dup(in int, out f1 int, out f2 text)
    AS $$ SELECT $1, CAST($1 AS text) || ' is text' $$
    LANGUAGE SQL;

SELECT * FROM dup(42);


CREATE TYPE dup_result AS (f1 int, f2 text);

CREATE FUNCTION dup(int) RETURNS dup_result
    AS $$ SELECT $1, CAST($1 AS text) || ' is text' $$
    LANGUAGE SQL;

SELECT * FROM dup(42);

CREATE FUNCTION dup(int) RETURNS TABLE(f1 int, f2 text)
    AS $$ SELECT $1, CAST($1 AS text) || ' is text' $$
    LANGUAGE SQL;

SELECT * FROM dup(42);

CREATE FUNCTION check_password(uname TEXT, pass TEXT)
RETURNS BOOLEAN AS $$
DECLARE passed BOOLEAN;
BEGIN
        SELECT  (pwd = $2) INTO passed
        FROM    pwds
        WHERE   username = $1;

        RETURN passed;
END;
$$  LANGUAGE plpgsql
    SECURITY DEFINER
    -- Set a secure search_path: trusted schema(s), then 'pg_temp'.
    SET search_path = admin, pg_temp;


BEGIN;
CREATE FUNCTION check_password(uname TEXT, pass TEXT)
RETURNS BOOLEAN AS $$
DECLARE passed BOOLEAN;
BEGIN
        SELECT  (pwd = $2) INTO passed
        FROM    pwds
        WHERE   username = $1;

        RETURN passed;
END;
$$  LANGUAGE plpgsql
    SECURITY DEFINER;
REVOKE ALL ON FUNCTION check_password(uname TEXT, pass TEXT) FROM PUBLIC;
GRANT EXECUTE ON FUNCTION check_password(uname TEXT, pass TEXT) TO admins;
COMMIT;
```

Reply:
These all pass with your code btw - I checked!

Reply:
Yeah, happy to add them. It exposed another problem though - the dollar_quote lexer is greedy. This regex: \$(\w*)\$[^\1]*\$\1\$ would match the entirety of this string "$$ function_stuff $$ stuff after function $$ another_function's stuff $$

Reply:
And that's why extra test cases are useful 😁

Agree on your regex change.

Though looks like it still doesn't like `in`/`out` params nor `RETURNS TABLE`  Both should be possible by overriding the appropriate pieces:

```py
     FunctionParameterGrammar=OneOf(
         Sequence(
             OneOf("IN", "OUT", optional=True),
             Ref("ParameterNameSegment", optional=True),
             OneOf(Sequence("ANY", "TYPE"), Ref("DatatypeSegment")),
         ),
```

Weirdly though I can't get it to recognise nameless In params (`in int`) like in the example, but named ones do work (`in a1 int`). Spot anything wrong with this?

```py
         Ref("FunctionParameterListGrammar"),
         Sequence(  # Optional function return type
             "RETURNS",
             OneOf(
                 Ref("DatatypeSegment"),
                 Sequence(
                     Ref.keyword("TABLE"),
                     Ref("FunctionParameterListGrammar"),
                 ),
             ),
             optional=True,
```

Reply:
Yeah, I've been having a fight with it for almost an hour. I've solved the vars problem, just can't quite get the table syntax right, The issue you're having is that DataTypeSegment isn't well-defined, its regex is basically anything, So it's finding the data type as the optional ParameterName, then looking after it for the required DataType. This code works, wrapping it in OneOf instead of using optional:

`
    FunctionParameterGrammar=Sequence(
        OneOf("IN", "OUT", "INOUT", "VARIADIC", optional=True),
        OneOf(Ref("DatatypeSegment"),
              Sequence(Ref("ParameterNameSegment"),
                       Ref("DatatypeSegment"))),
        Sequence(
            OneOf(
                "DEFAULT",
                Ref("EqualsSegment")
            ),
            Ref("LiteralGrammer"),
            optional=True
        )
    )
`

Reply:
Right, changes made, all tests passing locally... i think we might finally be there 😄